src/ring_theory/dedekind_domain/S_integer.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# `S`-integers and `S`-units of fraction fields of Dedekind domains
+
+Let `K` be the field of fractions of a Dedekind domain `R`, and let `S` be a set of prime ideals in
+the height one spectrum of `R`. An `S`-integer of `K` is defined to have `v`-adic valuation at most
+one for all primes ideals `v` away from `S`, whereas an `S`-unit of `Kˣ` is defined to have `v`-adic
+valuation exactly one for all prime ideals `v` away from `S`.
+
+This file defines the subring of `S`-integers of `K` and the subgroup of `S`-units of `Kˣ`, which
+can be specialised to the case of a number field or a function field separately.
+
+TODO:
+ * notation for `S`-integers and `S`-units.
+ * proof that `S`-units is the kernel of a map to a product.
+ * proof that `∅`-integers is the usual ring of integers.
+ * finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+ * `set.integer`: `S`-integers.
+ * `set.unit`: `S`-units.
+
+## Main statements
+
+ * `set.units_integer_equiv_unit`: units of `S`-integers are `S`-units.
+
+## References
+
+ * [D Marcus, *Number Fields*][marcus1977number]
+ * [J W S Cassels, A Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+ * [J Neukirch, *Algebraic Number Theory*][Neukirch1992]
+
+## Tags
+
+S integer, S-integer, S unit, S-unit
+-/
+
+open is_dedekind_domain
+
+open_locale non_zero_divisors
+
+namespace set
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R]
+  (S : set $ height_one_spectrum R) (K : Type v) [field K] [algebra R K] [is_fraction_ring R K]
+
+/-! ## `S`-integers -/
+
+/-- The subring of `S`-integers of `K`. -/
+@[simps] def integer : subring K :=
+{ carrier   := {x : K | ∀ v ∉ S, (v : height_one_spectrum R).valuation x ≤ 1},
+  mul_mem'  := λ _ _ hx hy v hv, by simp only [map_mul, mul_le_one₀ (hx v hv) (hy v hv)],
+  one_mem'  := λ _ _, by simp only [map_one, le_refl],
+  add_mem'  := λ _ _ hx hy v hv, v.valuation.map_add_le (hx v hv) (hy v hv),
+  zero_mem' := λ _ _, by simp only [map_zero, zero_le_one],
+  neg_mem'  := λ _ hx v hv, by simp only [valuation.map_neg, hx v hv] }
```

Comment:
```suggestion
(⨅ v ∉ S, (v : height_one_spectrum R).valuation.valuation_subring.to_subring).copy
  {x : K | ∀ v ∉ S, (v : height_one_spectrum R).valuation x ≤ 1}
  (set.ext $ λ _, by simpa only [set_like.mem_coe, subring.mem_infi])
```
It takes 4.44s for decl post-processing though, while the original one takes only takes 0.27s.